x-pack/plugins/lens/public/react_embeddable/expressions/on_event.ts

```diff
@@ -0,0 +1,113 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { ExpressionRendererEvent } from '@kbn/expressions-plugin/public';
+import { VIS_EVENT_TO_TRIGGER } from '@kbn/visualizations-plugin/public';
+import { type AggregateQuery, type Query, isOfAggregateQueryType } from '@kbn/es-query';
+import {
+  isLensBrushEvent,
+  isLensEditEvent,
+  isLensFilterEvent,
+  isLensMultiFilterEvent,
+  isLensTableRowContextMenuClickEvent,
+} from '../../types';
+import { inferTimeField } from '../../utils';
+import type {
+  GetStateType,
+  LensApi,
+  LensEmbeddableStartServices,
+  LensPublicCallbacks,
+} from '../types';
+import { isTextBasedLanguage } from '../helper';
+import { addLog } from '../logger';
+
+export const prepareEventHandler =
+  (
+    api: LensApi,
+    getState: GetStateType,
+    callbacks: LensPublicCallbacks,
+    { data, uiActions, visualizationMap }: LensEmbeddableStartServices,
+    disableTriggers: boolean | undefined
+  ) =>
+  async (event: ExpressionRendererEvent) => {
+    if (!uiActions?.getTrigger || disableTriggers) {
+      return;
+    }
+    addLog(`onEvent$`);
+
+    let eventHandler:
+      | LensPublicCallbacks['onBrushEnd']
+      | LensPublicCallbacks['onFilter']
+      | LensPublicCallbacks['onTableRowClick'];
+    let shouldExecuteDefaultTriggers = true;
+
+    if (isLensBrushEvent(event)) {
+      eventHandler = callbacks.onBrushEnd;
+    } else if (isLensFilterEvent(event) || isLensMultiFilterEvent(event)) {
+      eventHandler = callbacks.onFilter;
+    } else if (isLensTableRowContextMenuClickEvent(event)) {
+      eventHandler = callbacks.onTableRowClick;
+    }
+    const currentState = getState();
+    // if the embeddable is located in an app where there is the Unified search bar with the ES|QL editor, then use this query
+    // otherwise use the query from the saved object
+    let esqlQuery: AggregateQuery | Query | undefined;
+    if (isTextBasedLanguage(currentState)) {
+      const query = data.query.queryString.getQuery();
+      esqlQuery = isOfAggregateQueryType(query) ? query : currentState.attributes.state.query;
+    }
+
+    eventHandler?.({
+      ...event.data,
+      preventDefault: () => {
+        shouldExecuteDefaultTriggers = false;
+      },
+    });
+
+    if (isLensFilterEvent(event) || isLensMultiFilterEvent(event) || isLensBrushEvent(event)) {
+      if (shouldExecuteDefaultTriggers) {
```

Comment:
It looks like you are expecting a sync update of the variable `shouldExecuteDefaultTriggers` coming from the event handler, are we sure we can trust a sync call to preventDefault()?

Reply:
While not explicit, the documented example of usage for the `preventDefault` hints to call it in a sync way.
I could provide a better documentation direction as a followup.

Reply:
ok, seems a bit fragile anyway.